src/Path/Include.hs

```diff
@@ -334,47 +346,105 @@ dirname (Path "") = Path ""
 dirname (Path l) | FilePath.isDrive l = Path ""
 dirname (Path l) = Path (last (FilePath.splitPath l))
 
+-- | Split the given file path into a filename and an extension. If the
+-- filename has no extension it returns "" as extension.
+splitExtension :: Path b File -> (Path b File, String)
+splitExtension path@(Path fpath) =
+    if nameDot == [] || ext == [] || all FilePath.isExtSeparator nameDot
+    then (path, "")
+    else (Path (drv ++ dir ++ (init nameDot)), FilePath.extSeparator : ext)
+
+    where
+
+    -- trailing separators are ignored for the split and considered part of the
+    -- second component in the split.
+    splitLast isSep str =
+        let rstr = reverse str
+            notSep = not . isSep
+            name = (dropWhile notSep . dropWhile isSep) rstr
+            trailingSeps = takeWhile isSep rstr
+            xtn  = (takeWhile notSep . dropWhile isSep) rstr
+        in (reverse name, reverse xtn ++ trailingSeps)
+
+    (drv, pth)     = FilePath.splitDrive fpath
+    (dir, file)    = splitLast FilePath.isPathSeparator pth
+    (nameDot, ext) = splitLast FilePath.isExtSeparator file
+
 -- | Get extension from given file path.
 --
+-- >>> fileExtension $(mkRelFile "name.foo"     ) == ".foo"
+-- >>> fileExtension $(mkRelFile "name.foo."    ) == ".foo."
+-- >>> fileExtension $(mkRelFile "name.foo.."   ) == ".foo.."
```

Comment:
Dots in trailing position of an extension are valid, they can be semantically allowed without violating any rule. In fact they are necessary so that we can consistently and unambiguously compose/decompose extensions of this form:

```
name.foo...bar
```

The extension in this case is a composition of `.foo..` + `.bar`, there is no ambiguity. Notice that, if we drop the extension from `name.foo...bar` we will get `name.foo..` and it has an extension `.foo..`, everything is consistent.

Reply:
I allowed trailing dots in a separate commit, maybe you were looking at the first commit only where `.foo..` was invalid?